### DISCUSSION TITLE
Best practice for touch events (e.g. modeling swipe)

### DISCUSSION THREAD

**Discussion by efosler** [zauberzeug/nicegui#2759] [Q&A] [answered]
### Question

Hi,  currently playing around with NiceGUI and I was looking to add a touch-swipe style action for a `ui.carousel` (currently running on a touch screen attached to a Raspberry Pi).  Was able to capture the `touch*` events but the `touches` field in the field comes up empty. (Possible bug?)  Have confirmed using javascript that the browser does indeed provide the touches field.  

I started going down the rabbit hole of trying to either detect swipes myself, or to use the touch swipe directive within quasar (which seems like a more robust approach).  Any hints on how to add a touch swipe to a nicegui element like `ui.carousel`?  

NB: I started to implement the solution in [the Quasar touch directive page](https://quasar.dev/vue-directives/touch-swipe/ ) but noticed that when I tried to pass a property `v-touch-swipe.mouse="handleSwipe"`, the props function splits `v-touch-swipe` and `mouse` and I end up with a property `mouse="handleSwipe"`.

Anyway, rather than reinvent the wheel (which has been interesting, mind you), I thought I'd ask if there was a best practice for adding touch events or quasar touch directives to nicegui elements.

**Comment by falkoschindler** [comment, marked as answer]
Hi @efosler,

To add a "touch-swipe style action for a `ui.carousel`", I guess you can simply add the "swipeable" prop.

For more general touch events, you might have spotted a limitation in the `.props()` method. I wasn't aware of possible dots in the prop name. Currently it only supports `\w` (a-z, A-Z, 0-9 and _) and `-`:
https://github.com/zauberzeug/nicegui/blob/ee14497b1ee02c8f1dbff43964bc8d639704e116/nicegui/element.py#L26

We can probably simply add a dot. As a workaround, you can access the private props dictionary like this:
```py
carousel._props['v-touch-swipe.mouse'] = 'handleSwipe'
```

**Reply by efosler** [reply]
Thanks for the tips!   Relative newbie to nicegui/quasar/tailwind so I appreciate the help.

**Reply by efosler** [reply]
Just putting this here in case someone else runs into my issue: adding the "swipeable" prop seemed to help with the chromium browser but I'm still experiencing issues with the firefox browser.  This seems to be an issue with quasar/firefox at least for my setup (I can't get the quasar touch examples to run under firefox).  Starting a discussion over there.

**Reply by falkoschindler** [reply]
https://github.com/quasarframework/quasar/discussions/17064

**Comment by HeiziND** [comment]
Hi, I sorry to hijack this old thread but it's the only place where I found related discussion to my problem.

I am trying to use v-touch-pan to implement a resizer for my drawer, largely taking inspiration from this : 
https://codepen.io/pdanpdan/pen/mdLjrGL

Notwithstanding the problem with the nicegui not handling props with dots, I am unable to get even the simplest vue directives to fire in my nicegui app, I tried both Chrome and Safari. 

here's a minimal example to reproduce the issue - nothing is logged to the javascript console in my case. I tried v-touch-swipe , v-touch-pan, etc and none works.
I am *very* new to nicegui and web/uis so it could be very well be something silly that I a missing. Appreciate any pointers to the right direction 🙏.

```python
@ui.page('/vue_directive')
async def page_vue_directive():
    ui.add_body_html('''
<script>
  function onPan(event) {
    console.log(event)
  }
  // Make the function globally accessible
  // window.onPan = onPan;
</script>
    ''')
    await ui.context.client.connected()
    with ui.element('div').props('v-touch-pan="onPan"').style('width: 100px; height: 100px; background-color: red;'):
        ui.label('pan me')
 ```

**Reply by falkoschindler** [reply]
@HeiziND To be honest, I also don't know how to use such directives. Just recently I tried calling a global function for quite a while, but without any luck: #3754. Maybe someone from the community can help? But to get more attention, I suggest to create a new discussion.

**Reply by HeiziND** [reply]
@falkoschindler , if the man himself can't get it to work than what hope do I - a mere mortal - stand :)
joking a aside, I think I will drop this quest for now and try to find simpler solutions for my resizing needs. I am very new to nicegui and should probably not start fighting the framework (yet) but rather adapt to what it can do out of the box.
